### DISCUSSION TITLE
How to catch (hook into) a soft-reset, to deinitialize a library/driver?

### DISCUSSION THREAD

**Discussion by moefear85** [micropython/micropython#12850] [Libraries & Drivers] [answered]
Hi,

Some drivers (User C Modules on esp32) need to be deinitialized before a soft-reset, like the esp32-camera and espnow, otherwise they (can) crash the system. I can't find any example on how to get a soft reset (or ctrl+D) to automatically trigger my deinit() function. Is this possible? I need it for [this](https://github.com/moefear85/micropython_with_espnow_camera) project.

**Comment by 2dof** [comment]
before  calling ```sys.exit()``` (A soft reset simply clears the state of the MicroPython virtual machine)   run your ```deinit() ```function , then wait some time (to give time to 'shut-down' camera driver ) and then exit() .

edit, You can also pass function in  sys.exit(deinit()):
```
def deinit():
    print('soft reset') 
    #  do driver's shut-down ....

sys.exit(deinit())

```
 
note that there is also a machine.soft_reset()  (Performs a soft reset of the interpreter, deleting all Python objects and resetting the Python heap) , but it do not accept any input parameters.

read more: [Reset and boot modes](https://docs.micropython.org/en/latest/wipy/tutorial/reset.html)

**Reply by moefear85** [reply]
sys.exit() doesn't trigger a soft reset right? I need deinit() to get called automatically when the user presses ctrl+D manually, rather than a procedural exit. How do other drivers handle this? Don't they get corrupted when C parts continue running while python parts get all deleted, such as callbacks?

**Reply by dlech** [reply, marked as answer]
Answering the question how do other drivers do this: For things like file handles, since the types are implemented in C, they can use a finalizer (assuming the port you are using has GC finalization enabled) to release the resources. There is a special GC function that is called in most ports that ensures that all finalizers are run on soft reset.

But if all of your code is written in Python, I'm not sure what could be done.

**Reply by robert-hh** [reply]
If present in a port, machine.soft_reset() causes a soft reset. Ctr-D will only call a soft reset if at the REPL prompt. When executing user code, Ctrl-D is just a byte delivered by sys.stdin.read(), or it terminates a call to input().

**Reply by moefear85** [reply]
thx, for the hints. I'll research finalizers. I think they're available on the esp port. I forgot to mention the driver is a C module.

**Comment by shariltumin** [comment]
May I suggest this. You will need to modify the 'main.c' file. As you are making a custom firmware, this will not be a problem for you.

In 'main.c', add a single new line at line 153:

```c
soft_reset_exit:

    int sret = pyexec_file_if_exists("sreset.py");

```

Now upload 'sreset.py' to the flash.

As an example, 'sreset.py' might have the following lines:


```py
import time

print('doing soft reset')
print('deinit all the drivers here')
print('....')
print('wait a while')
time.sleep(3)
```

Time to see if it works

```
MicroPython v1.22.0-preview.63.g2fda94c28.dirty on 2023-11-02; ESP32 ACT/BLE KAKI5 with ESP32
>>> os.listdir()
['boot.py', 'sreset.py']
>>> Ctrl+d
doing soft reset
deinit all the drivers here
....
wait a while
MPY: soft reboot
MicroPython v1.22.0-preview.63.g2fda94c28.dirty on 2023-11-02; ESP32 ACT/BLE KAKI5 with ESP32
>>>
```
It works!

**Reply by moefear85** [reply]
oh wow thx... this seems like an inofficial way to do things, but is quicker than having to figure out finalizers so I can continue working immediately. thx everyone for the help. I'll be sharing the source and images when I'm done.
